pkg/talos/nodeconfig.go

```diff
@@ -126,6 +138,106 @@ func applyNodeOverride(node *config.Node, cfg taloscfg.Provider) taloscfg.Provid
 	return cfg
 }
 
+// Template supported fields within the config
+func templateConfig(node *config.Node, cfg taloscfg.Provider) (taloscfg.Provider, error) {
+	// Two separate copies of the configuration are required: one that is
+	// continuously updated with rendered values, and one that that contains
+	// the constant, raw template values. This ensures consistent results in
+	// the event that support for new fields are added, or that they are
+	// templated in a different order, which is important to prevent breaking
+	// changes.
+	renderedConfig := cfg.RawV1Alpha1()
+	unchangedConfig := renderedConfig.DeepCopy()
+
+	err := templateConfigField(node.NodeLabels, &renderedConfig.MachineConfig.MachineNodeLabels, unchangedConfig, "node labels")
+	if err != nil {
+		return nil, err
+	}
+
+	err = templateConfigField(node.NodeAnnotations, &renderedConfig.MachineConfig.MachineNodeAnnotations, unchangedConfig, "node annotations")
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
```

Comment:
This could alternatively be written as:

```go
func templateConfig(node *config.Node, cfg taloscfg.Provider) (taloscfg.Provider, error) {
	// Two separate copies of the configuration are required: one that is
	// continuously updated with rendered values, and one that that contains
	// the constant, raw template values. This ensures consistent results in
	// the event that support for new fields are added, or that they are
	// templated in a different order, which is important to prevent breaking
	// changes.
	renderedConfig := cfg.RawV1Alpha1()
	unchangedConfig := renderedConfig.DeepCopy()

	fields := []struct{
		name string
		srcKeyPairs map[string]string
		dstKeyPairs *map[string]string
	}{
		{
			name: "node labels",
			srcKeyPairs: node.NodeLabels,
			dstKeyPairs: &renderedConfig.MachineConfig.MachineNodeLabels,
		},
		{
			name: "node annotations",
			srcKeyPairs: node.NodeAnnotations,
			dstKeyPairs: &renderedConfig.MachineConfig.MachineEnv,
		},
	}

	for _, field := range fields {
		err := templateConfigField(field.srcKeyPairs, field.dstKeyPairs, unchangedConfig, field.name)
		if err != nil {
			return nil, err
		}
	}

	return cfg, nil
}
```

The benefit here is that it's _extremely_ simple to add support for more fields - just add a new struct instance and you're done. The downside is some more complexity and abstraction. Do you have a preference?

Reply:
I'm fine with both actually. There might be other types outside of `map[string]string` that need to be templated right? The alternative seems to be easier to work with converting type to `map[string]string`?

Reply:
> There might be other types outside of map[string]string that need to be templated right?

I think this can be broken down into two parts: the value to be templated, and it's structural relationship with other data.

Templates must always be strings, but they could be converted as appropriate to other destination types. For example a template to calculate `bool` would always be a `string`, but could produce a truthy string (`true`/`false`) that should be converted to a `bool`.

Due to the nature of the config, each value (templated or not) must always be associated with a string-type key. In some cases (arrays), multiple values will be associated with a single key. 

With this in mind, the source type for a block of data will always be (after flattening/moving a "bottom" node in the config tree) either `map[string]string`, or `map[string][]string`. However, the destination can only be as specific as `map[string]any` and `map[string][]any`, as the resultant type may need to be something other than a string.

I think all cases could be covered by:
* Changing the destination type from `map[string]string` to map[string]any`, then adding some logic to convert the rendered value to each resultant literal type (rejecting `interface{}` types with an unsupported error)
* Adding similar logic for `map[string][]string`/`map[string][]any`

Does this approach work? If so, I'll pull all (or most) of the templating code into a separate package and then start working on this.

Reply:
Thanks! Just do what you seem fit. I need to see the code to know how that works because I'm bad at understanding words :D

Reply:
I worked on this quite a bit and was unable to find a way to generally describe the destination key pair types due to language constraints. I was able to make a couple of functions more generic (so that rendered types other than `string` are supported), and also added functions for templating slices.